includes/Jobs/GenerateProductFeed.php

```diff
@@ -18,18 +19,84 @@ class GenerateProductFeed extends AbstractChainedJob {
 
 	use BatchQueryOffset, LoggingTrait;
 
+	/**
+	 * Array of processing results.
+	 *
+	 * @var array $processed_items.
+	 */
+	protected $processed_items = array();
+
+	/**
+	 * Feed file creation and manipulation utility.
+	 *
+	 * @var FacebookFeedFileHandler $feed_file_handler.
+	 */
+	protected $feed_file_handler;
+
+	/**
+	 * Feed file creation and manipulation utility.
+	 *
+	 * @var FeedDataExporter $feed_data_exporter.
+	 */
+	protected $feed_data_exporter;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param ActionSchedulerInterface $action_scheduler   Action Scheduler facade.
+	 * @param FacebookFeedFileHandler  $feed_file_handler  Feed file creation and manipulation handler.
+	 * @param FeedDataExporter         $feed_data_exporter Handling of file data.
+	 */
+	public function __construct( ActionSchedulerInterface $action_scheduler, $feed_file_handler, $feed_data_exporter ) {
+		parent::__construct( $action_scheduler );
+		$this->feed_file_handler  = $feed_file_handler;
+		$this->feed_data_exporter = $feed_data_exporter;
+	}
+
 	/**
 	 * Called before starting the job.
 	 */
 	protected function handle_start() {
-		// Optionally override this method in child class.
+		$this->feed_file_handler->prepare_feed_folder();
+		$this->feed_file_handler->create_fresh_feed_temporary_file();
+		$this->feed_file_handler->write_to_feed_temporary_file(
+			$this->feed_data_exporter->generate_header()
+		);
 	}
 
 	/**
 	 * Called after the finishing the job.
 	 */
 	protected function handle_end() {
-		// Optionally override this method in child class.
+		$this->feed_file_handler->replace_feed_file_with_temp_file();
+	}
+
+	/**
+	 * Handle processing a chain batch.
+	 *
+	 * @hooked {plugin_name}/jobs/{job_name}/chain_batch
+	 *
+	 * @param int   $batch_number The batch number for the new batch.
+	 * @param array $args         The args for the job.
+	 *
+	 * @throws Exception On error. The failure will be logged by Action Scheduler and the job chain will stop.
+	 */
+	public function handle_batch_action( int $batch_number, array $args ) {
+		$this->processed_items = array();
+		parent::handle_batch_action( $batch_number, $args );
+		$this->write_processed_items_to_feed();
```

Comment:
Because the framework ( correctly - because 99% this is what someone would want ) processes the items one by one but for writing reasons we need to write in one go. This behavior was triggered by a comment we got from VIP https://github.com/woocommerce/facebook-for-woocommerce/issues/1901#issuecomment-837687668.

@danielbitzer Could we make some changes to the framework to maybe have `handle_batch_action_start` and 'handle_batch_action_end` functions that would allow making this cleaner? Or this is good enough?

Reply:
Aha, thanks for explaining. Since this framework is tailored to our needs, would be good to align the interface to what we need. The main thing I'm concerned with here is making it clear where the state accumulates over each item, and when it's cleared out - feels like a "global variable" pattern so just want to make it super clear.

Reply:
Interesting! I'd be ok with this solution but I'm keen to tweak the framework if it makes is easier for us to develop with it. I [opened a PR](https://github.com/woocommerce/action-scheduler-job-framework/pull/2) to add a `::process_items()` method.

If we added this I think we could something like this:

```php

	/**
	 * Processes a batch of items.
	 *
	 * @since 1.1.0
	 *
	 * @param array $items The items of the current batch.
	 * @param array $args  The args for the job.
	 *
	 * @throws Exception On error. The failure will be logged by Action Scheduler and the job chain will stop.
	 */
	protected function process_items( array $items, array $args ) {
		$processed_items = array();

		foreach ( $items as $item ) {
			$processed_items[] = $this->process_item( $item, $args );
		}

		$this->feed_file_handler->write_to_feed_temporary_file(
			$this->feed_data_exporter->format_items_for_feed( $this->processed_items )
		);
	}




```

Reply:
:shipit:

Reply:
Implemented.